### PR TITLE
the Strike Team Revamp

### PR DESCRIPTION
**Pull request by JMWTurner**
It's happening!

Here's how strike team work now:
* When a strike team is called, a message is sent to every ghost that isn't job-banned from Strike Teams. After a minute, if at least one ghost applied, every selected applicant gets spawned at their spawning landmarks.
* If there are enough commando, one of them might be a leader. Other commando have to follow his orders. The deathsquad leader (Creed) can spawn when there are at least 4 deathsquads, the ERT leader always spawn even if he's alone, the Elite Syndicate strike teams don't have leaders.

The revamp isn't fully over though, next come the map changes. I'll wait for this part to go through.

secret
save
recruit
bravo

:cl:
* rscadd: Added strike team datums, thus purging lots of copypasted code, and allowing for the easy adding of new strike teams types.
* tweak: Removed the JoinResponseTeam and the like buttons. Now ghosts join strike teams just like dionae and posibrains.
* rscadd: The strike team buttons for admins have been grouped together for easy use.
* experimental: Added framework for custom strike teams.
* tweak: Syndicate Elite Strike Teams now spawn with OSIPR instead of pulse rifles. Although they still cannot reach the station on most maps yet.

### CODE REVIEW

**Comment by jakmak5** [comment]
Maybe also lower the access requirment on the comms console from captain level to instead normal head of staff level. And maybe include calling ERT via head of staff card swippers aswell for a redundant way of calling ERT in emergency incase the comms consoles cant be accessed.

**Comment by Aurx** [comment]
Creed is supposed to be the admin, not a PC.  
And a corpse already spawns in the office for easy DMPCing, doesn't it?

**Comment by JMWTurner** [comment]
>Creed is supposed to be the admin, not a PC.

@Aurx 
One of the points of this revamp is to build the framework for allowing ERTs to operate fully without admin control. In the future, the ERT commander could send a deathsquad request to Nanotrasen for example, so players would do their briefing by themselves using the report sent from the station.

I've never ever seen an admin impersonate Creed and the Seraph has been collecting dust for years. I'd like to shuffle things around a bit, and I'm sure it's gonna be lots of fun for the players. It reminds me of how you get a chance to become a jedi in Battelfront 2.

TL;DR: Let's allow players to become Creed, and lead the Deathsquad into action.

>And maybe include calling ERT via head of staff card swippers

@jakmak5 
That's....been a thing since December 5th. Comic added it.

**Comment by Aurx** [comment]
If it's something planned for a future PR, why are you putting aspects of it into a current PR?  

I don't like the idea of fully player-initiated ERTs.  The ability to add a group of geared-up players is one that can utterly reshape a round, and I don't think that moving that from an admin option to a player option will make things more fun.  Team antags and blobs will be screwed when they have to deal with a on-demand killsquad, and they shouldn't have to stay secret or deny comms console access just to not deal with centcomm.  
A handful of assholes who know the roundtype, have AEGuns, and have no actual job to keep them busy once they resolve whatever issue they were called for (assuming it wasn't "Kill the antags") is also a pretty easy way to ruin a round.  

No player should ever have the ability to summon a deathsquad or get deathsquad gear into the round by any means.  That needs to remain a strictly administrative decision.  

The deathsquad is automatically assigned ranks and a leader already.  

>I've never ever seen an admin impersonate Creed and the Seraph has been collecting dust for years. 

Creed and the Seraph collects dust because the deathsquad area and mechs collect dust because deathsquads are very rarely used.  I mean, check the logs and see how far back you have to go to find the last deathsquad.  It'll likely be months.

**Comment by JMWTurner** [comment]
>If it's something planned for a future PR, why are you putting aspects of it into a current PR?

@Aurx Because PRs have to be atomic, and so we can make sure what's been already coded is working correctly before adding stuff on top.

>I don't like the idea of fully player-initiated ERTs

Er...Aurx...fully player-initiated ERTs have been in the game for quite some time now. I think you mean player-initiated deathsquad, and that's not part of this PR.

In the near future, I'd like ERT commanders to be able to request a deathsquad but at first it'll still require admins to allow it.

**Comment by Aurx** [comment]
Must have missed that change then.  Still don't like it, but apparently it hasn't been ruining rounds so oh well.

**Comment by SonixApache** [comment]
The biggest issue with ERT right now is that they can't get back to centcomm once their mission is complete.

**Comment by Aurx** [comment]
I can't see ERT willingly going back to centcomm (and removing themselves from the round, effectively) except under duress.

**Comment by JMWTurner** [comment]
@SonixApache @Aurx In the mapping part of the revamp (that'll come in a PR following this one, the ERT will gain access to new rooms at centcom once they come back, their mission accomplished. Rooms featuring fun weapons, items, and maybe features I'll implement then. That way, ERT earns the right to access those spoils by doing their job first. As for how that is determined, I'm thinking of having the commander spawn with a paper containing their mission order, and have them ask the station captain to stamp it. On one hand, it means that a good ERT will make sure the station has a captain in control before they leave, on the other hand it means ERT can cheat by stealing the stamp, which I'm okay with as long as there's proper escalation. Otherwise, players can now be jobbanned from all strike teams.

I spoke above about how the ERT commander could request a deathsquid in the future, but more in general I want to increase the commander's role as a direct hotline between the station and centcom/nanotrasen. As if he had an integrated comms computer, allowing him to call the shuttle if there are no comms computer on the bridge, send IC messages to the admins asking for directions, etc...

Another idea would be to allow ERTs who complete their mission and come back to join the station as randomized name/body assistants. Those are just ideas for future PRs, This PR makes it so if we ever code something like that, we can easily have it work with Deathsquad and other strike teams without having to snowflake our way through new procs and stuff.

**Comment by SonixApache** [comment]
I very much like the random name/body assistant idea.

**Comment by Aurx** [comment]
Most players, if given a choice between being functionally dead but having a shiny toy they can play with on Z2 or staying alive will only choose the former once if ever.  
Additionally, players should not be on Z2 outside controlled inventory situations (IE, they spawn there without anything that could cause trouble), the emergency shuttle/pods, or admin interference.  Returning to cent with whateverthefuck in their pack is just begging for trouble.

**Comment by Probe1** [comment]
You could just delete them on returning to centcom and immediately send them back to the lobby screen.

Observers get a special treat of rejoining a round and emergency responders get motivation to leave the station once the emergency has passed.

**Comment by JMWTurner** [comment]
I like that there's discussion about it but none of that is in this PR. I'm taking notes but let's wait until we get to the next PR before deciding what to do with returning Strike Teams.

**Comment by clusterfack** [comment]
@JMWTurner I don't think you're thinking about the peak QoL that should be possible for admins when making this. Something like this would be far more useful in my opinion:
usefulshit
Simple tools to create a team, add objectives, modify them, recruit ghosts to them, monitor them, dissolve them, kill them, etc. All of this would be possible with some very simple functions and a datum to store all the information. Stylistically I have no problem with what you're showing in your screenshots, but functionally it seems a bit wanting?

(Also I think Creed is not meant to be a PC but an admin meta character.)

File: code/game/machinery/computer/communications.dm [hunk 1]
```diff
@@ -180,13 +180,18 @@ var/shuttle_call/shuttle_calls[0]
 				to_chat(usr, "<span class='notice'>\The [src.name] cannot establish a bluespace connection.</span>")
 				return
 
-			if(sent_strike_team)
+			if("Death Squad" in sent_strike_teams)
```
Comment (PJB3005) [inline code/game/machinery/computer/communications.dm:183]: This should be made a helper and at the very least use defines.

File: code/game/machinery/computer/communications.dm [hunk 2]
```diff
@@ -180,13 +180,18 @@ var/shuttle_call/shuttle_calls[0]
...
 				to_chat(usr, "<span class='warning'>PKI AUTH ERROR: SERVER REPORTS BLACKLISTED COMMUNICATION KEY PLEASE CONTACT SERVICE TECHNICIAN</span>")
 				return
 
-			if(world.time < 6000)
-				to_chat(usr, "<span class='notice'>The emergency response team is away on another mission, Please wait another [round((6000-world.time)/600)] minute\s before trying again.</span>")
+			if("Emergency Response Team" in sent_strike_teams)
```
Comment (PJB3005) [inline code/game/machinery/computer/communications.dm:187]: Same thing.

File: code/game/machinery/computer/communications.dm [hunk 3]
```diff
@@ -393,7 +397,11 @@ var/shuttle_call/shuttle_calls[0]
 		list("id"=SEC_LEVEL_BLUE,  "name"="Blue"),
 		//SEC_LEVEL_RED = list("name"="Red"),
 	)
-	data["ert_sent"] = send_emergency_team
+	var/sent = 0
+	if ("Emergency Response Team" in sent_strike_teams)
```
Comment (PJB3005) [inline code/game/machinery/computer/communications.dm:401]: Again.

File: code/game/striketeams/striketeam_datums.dm [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	var/list/applicants = list()
+	var/searching = FALSE
+
+	var/leader_key = ""
+	var/list/team_composition = list()
+
+
+/datum/striketeam/proc/trigger_strike(var/mob/user)
+	//Is the game started
+	if(!ticker)
```
Comment (PJB3005) [inline code/game/striketeams/striketeam_datums.dm:22]: Ticker existing does not necessarily entail the game having started.

The ticker exists as soon as the MC begins set up, finishes pre-round init when the MC is done, but this is all before the lobby timer even begins to tick down.

File: code/game/striketeams/striketeam_datums.dm [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+		if(user)
+			to_chat(user, "<span class='warning'>The game hasn't started yet!</span>")
+		qdel(src)
+		return
+
+	//Has someone already sent a strike team of this type
+	if(striketeam_name in sent_strike_teams)
```
Comment (PJB3005) [inline code/game/striketeams/striketeam_datums.dm:29]: Helper here.

File: code/game/striketeams/striketeam_datums.dm [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+		mission = input(user, "Please specify which mission the [striketeam_name] shall undertake.", "Specify Mission", "")
+
+		if(!mission)
+			if(alert("Error, no mission set. Do you want to exit the setup process?",,"Yes","No")=="Yes")
+				qdel(src)
+				return
+			else
+				mission = initial(mission)
+
+		if(striketeam_name in sent_strike_teams)
```
Comment (PJB3005) [inline code/game/striketeams/striketeam_datums.dm:52]: Helper here.

File: code/game/striketeams/striketeam_datums.dm [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+			to_chat(user, "Looks like someone beat you to it.")
+			qdel(src)
+			return
+
+	sent_strike_teams |= striketeam_name
```
Comment (PJB3005) [inline code/game/striketeams/striketeam_datums.dm:57]: This line is redundant, doing `list[x] = thing` adds `x` to the list already.

File: code/game/striketeams/striketeam_datums.dm [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+	searching = TRUE
+
+	var/icon/team_logo = icon('icons/mob/mob.dmi', logo)
+	for(var/mob/dead/observer/O in dead_mob_list)
+		if(jobban_isbanned(O, "Strike Team") || O.client.is_afk())
+			continue
+
+		to_chat(O, "[bicon(team_logo)]<span class='recruit'>[faction_name] needs YOU to become part of its upcoming [striketeam_name]. (<a href='?src=\ref[src];signup=\ref[O]'>Apply now!</a>)</span>[bicon(team_logo)]")
+
+	spawn(600)
```
Comment (PJB3005) [inline code/game/striketeams/striketeam_datums.dm:72]: `1 MINUTES`

File: code/game/striketeams/striketeam_datums.dm [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+		searching = FALSE
+
+		for(var/mob/dead/observer/O in dead_mob_list)
+			if(jobban_isbanned(O, "Strike Team") || O.client.is_afk())
+				continue
+			to_chat(O, "[bicon(team_logo)]<span class='recruit'>Applications for [faction_name]'s [striketeam_name] are now closed.</span>[bicon(team_logo)]")
+
+		if(!applicants || applicants.len <= 0)
+			if(user)
+				to_chat(user, "<span class='warning'>[faction_name] received no applications for the [striketeam_name].</span>")
```
Comment (PJB3005) [inline code/game/striketeams/striketeam_datums.dm:82]: This should probably be logged to all admins because if the admin starting this logs out, the error goes on silent ears.

Also using warning which is used a lot IC seems like too easily missed.

File: code/game/striketeams/striketeam_datums.dm [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+			if(applicants.len <= 0)
+				break
+
+			var/mob/applicant = null
+			for(var/mob/M in player_list)
+				if(M.key == pick(applicants))
+					applicant = M
+			applicants -= applicant.key
+
+			if(!istype(applicant, /mob/dead/observer))
```
Comment (PJB3005) [inline code/game/striketeams/striketeam_datums.dm:108]: `isobserver()`

File: code/game/striketeams/striketeam_datums.dm [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+	var/mob/living/carbon/human/new_commando = new(spawn_location.loc)
+	return new_commando
+
+/datum/striketeam/proc/greet_commando(var/mob/living/carbon/human/H)
+	to_chat(H, "<span class='notice'>You are a [striketeam_name] commando, in the service of [faction_name].</span>")
+	to_chat(H, "<span class='notice'>Your current mission is: <span class='danger'>[mission]</span></span>")
+
+/datum/striketeam/Topic(var/href, var/list/href_list)
+	if(href_list["signup"])
+		var/mob/dead/observer/O = locate(href_list["signup"])
```
Comment (PJB3005) [inline code/game/striketeams/striketeam_datums.dm:139]: You shouldn't use the href to locate the observer here, go by `usr` instead.

Right now this would allow you to make anybody sign up if you knew the ref of your fellow ghosts.

File: code/game/striketeams/striketeam_datums.dm [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+/datum/striketeam/proc/failure()
+
+/datum/striketeam/proc/extras()
+
+/datum/striketeam/proc/volunteer(var/mob/dead/observer/O)
+	if(!searching || !istype(O))
+		return
+
+	if(jobban_isbanned(O, "Strike Team"))
+		to_chat(O, "<span class='warning'>Your application to the [striketeam_name] has been discarded due to past conduct..</span>")
```
Comment (PJB3005) [inline code/game/striketeams/striketeam_datums.dm:154]: While I think this flavor text is nice and all, being more descriptive would probably be a good idea.

File: code/game/striketeams/striketeam_datums.dm [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+		return
+
+	if(O in applicants)
```
Comment (PJB3005) [inline code/game/striketeams/striketeam_datums.dm:157]: Does this even work? The `applicants` list is clearly a list of keys here, doing `O in applications` would always be false, no?

I think you want to do `O.key in applications` here.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Comment by JMWTurner** [comment]
>(Also I think Creed is not meant to be a PC but an admin meta **character.)

@clusterfack we discussed about that point above with Aurx. In my opinion there's no reason to keep that character as meta-admin only. I'd love to see him on the battlefield, and in the hands of players.

Also, this PR adds strike team datums so what you're suggesting can indeed be easily added, but still I'd like to add those features after this part gets merged.

**Comment by clusterfack** [comment]
@JMWTurner That's silly he isn't meant to be a character that is in the thick of fighting, but a commander.

**Comment by clusterfack** [comment]
Also to be perfectly honest I don't see where the improvement is in this PR in the first place from the description. It does next to nothing.

**Comment by PJB3005** [comment]
@clusterfack it cleans up the strike team code, ghost recruiting and allows for easier bus.

**Comment by clusterfack** [comment]
It really doesn't seem to be significantly cleaning up strike team code more just moving it around than anything, and it seems like he's trying to sneak in changes by putting in a ton of changes at once so that they all cannot be noticed.

File: code/game/striketeams/syndicate_elite_squad.dm [hunk 14]
```diff
@@ -126,52 +38,59 @@ var/global/sent_syndicate_strike_team = 0
...
+		equip_to_slot_or_del(new /obj/item/weapon/plastique(src), slot_in_backpack)
+		equip_to_slot_or_del(new /obj/item/energy_magazine/osipr(src), slot_in_backpack)
 
-	equip_to_slot_or_del(new /obj/item/weapon/melee/energy/sword(src), slot_l_store)
-	equip_to_slot_or_del(new /obj/item/weapon/grenade/empgrenade(src), slot_r_store)
-	equip_to_slot_or_del(new /obj/item/weapon/tank/emergency_oxygen(src), slot_s_store)
-	equip_to_slot_or_del(new /obj/item/weapon/gun/projectile/silenced(src), slot_belt)
 
-	put_in_hands(new /obj/item/weapon/gun/energy/pulse_rifle(src)) //Will change to something different at a later time -- Superxpdude
+	put_in_hands(new /obj/item/weapon/gun/osipr(src))
```
Comment (clusterfack) [inline code/game/striketeams/syndicate_elite_squad.dm]: For example, this change

Comment (JMWTurner) [reply code/game/striketeams/syndicate_elite_squad.dm:93]: Syndicate elite strike teams still cannot be properly deployed on the station. And they won't be until I've sprited their new armors and re-mapped their shuttle and stuff. When they're once again operational, I'll specify all their new stuff in the changelog.

Comment (JMWTurner) [reply code/game/striketeams/syndicate_elite_squad.dm:93]: ps: these gun were meant to be the Elite Syndicate Strike Team's guns ever since they were conceived. https://github.com/d3athrow/vgstation13/blob/Bleeding-Edge/code/modules/projectiles/guns/projectile/osipr.dm#L7

I never bothered adding them there since their strike team doesn't work, and this PR means I'll be able to make them work in another PR without having to snowflake my way through.

**Comment by JMWTurner** [comment]
>more just moving it around than anything

@clusterfack It's both a revamp and a rewrite. You do realize that all 3 strike teams we currently have are snowflakes independent of each others right?

If my intention was to sneak in stuff I'd have done it in this PR: #13036

File: code/game/striketeams/syndicate_elite_squad.dm (hunk 14 above)
Comment (Exxion) [reply code/game/striketeams/syndicate_elite_squad.dm:93]: That reference may be a BIT too overt

Comment (JMWTurner) [reply code/game/striketeams/syndicate_elite_squad.dm:93]: Well we talked about it back in the days and no one had any issues with it at the time.

**Comment by clusterfack** [comment]
It would have been easy to see in that PR. Instead you made changes to a much larger PR that changes many lines without telling anyone what those changes were.

**Comment by JMWTurner** [comment]
You're right, I've made an entry in the changelog for those.

**Comment by Aurx** [comment]
>the syndithing that's never used now starts with immulsion breaking weapons

I oppose making OSIPRs part of any loadout ever, unless they get renamed to something less egregious.

File: code/game/striketeams/syndicate_elite_squad.dm (hunk 14 above)
Comment (Exxion) [reply code/game/striketeams/syndicate_elite_squad.dm:93]: The only way I'd believe that is if they didn't read that description
I don't have a problem with the gun existing
But that description is just too much

**Comment by JMWTurner** [comment]
@Aurx But the XCOM guns and armors we've had for years are fine? Just asking.

I'll rename it to BPR (ballistic pulse rifle) if you wish.

**Comment by Exxion** [comment]
They're a *bit* less overt
I mean
Not in the code
But in-game they are

**Comment by Aurx** [comment]
They don't outright shove "HI I'M FROM ANOTHER GAME REMEMBER YOU'RE PLAYING A GAME" in your face.

**Comment by JMWTurner** [comment]
>SINCE IT'S POPULAR RIGHT NOW

@Aurx er....you do realize Half Life 2 came out in 2004 right? It's not as popular as it was back then.

http://half-life.wikia.com/wiki/Overwatch_Standard_Issue_Pulse_Rifle

pls don't tell me you thought that the "overwatch" in its name stood for Blizzard's game, cause that gun's been in our code long before it even came out.

**Comment by Aurx** [comment]
I stand corrected then.  I first saw the OSIPR around that time and made an assumption.

**Comment by Exxion** [comment]
It looks like the one from HL2
Uses the sounds from HL2
References HL2 blatantly in its description
And you thought it was an Overwatch reference?

**Comment by Aurx** [comment]
The only 3d FPSes I play are Unreal and Tournament.

**Comment by Probe1** [comment]
Come back and fix this up!

**Comment by JMWTurner** [comment]
I'll do it next week, since I won't have class that week.

**Comment by JMWTurner** [comment]
No class but an assignment... I swear I'll get to it eventually

**Comment by Yelier** [comment]
Does this involve players being able to control the ERT shuttle? If so, that's not a good idea. Unless you make the ERT setup area completely cut off from the rest of Centcomm, it wouldn't be hard to break into the rest of the station. I don't think that the shutters over the doors are invincible.

**Comment by JMWTurner** [comment]
>Does this involve players being able to control the ERT shuttle?

@Yelier no, only ERT have access. Like they currently do.

File: code/game/striketeams/striketeam_datums.dm (hunk 4 above)
Comment (JMWTurner) [reply code/game/striketeams/striketeam_datums.dm:22]: I assumed it was just to prevent some bugs. Not like admins would be stupid enough to call a strike team before roundstart right?....right?

**Comment by JMWTurner** [comment]
Got rid of conflicts, now to add helper datums

**Comment by JMWTurner** [comment]
Moving to another PR because of git

We've moved at #14446
